nntrainer/dataset/databuffer_factory.cpp

```diff
@@ -40,33 +46,42 @@ std::unique_ptr<DataBuffer> createDataBuffer(DatasetType type,
   NNTR_THROW_IF(file == nullptr, std::invalid_argument)
     << "file shall not be null, use empty constructor instead";
 
+  std::unique_ptr<DataProducer> dp;
   switch (type) {
   case DatasetType::FILE:
-    return std::make_unique<DataBufferFromDataFile>(file);
+    dp = std::make_unique<RawFileDataProducer>(file);
+    break;
   case DatasetType::UNKNOWN:
     [[fallthrough]];
   default:
     throw std::invalid_argument(
       "Unsupported constructor type for the dataset of type: " +
       static_cast<int>(type));
   };
+
+  return std::make_unique<DataBuffer>(std::move(dp));
 }
 
 /**
  * @brief Factory creator with constructor for dataset
  */
 std::unique_ptr<DataBuffer> createDataBuffer(DatasetType type, datagen_cb cb,
                                              void *user_data) {
+
+  std::unique_ptr<DataProducer> dp;
   switch (type) {
   case DatasetType::GENERATOR:
-    return std::make_unique<DataBufferFromCallback>(cb, user_data);
+    dp = std::make_unique<FuncDataProducer>(cb, user_data);
+    break;
   case DatasetType::UNKNOWN:
```

Comment:
For a later PR, check if this enum is needed anymore.

Reply:
I guess we can't delete this as it's already part of capi.

Reply:
I cant find it in the capi. Can you point me where?

Reply:
My bad, we don't have it. As other enums does not have this, I thought we have this. I'll put this in #303